### PR TITLE
expression: fix misuse of datumsToConstants in GBK test

### PR DESCRIPTION
**Pull request by unconsolable**
Signed-off-by: unconsolable <chenzhipeng2012@gmail.com>

### What problem does this PR solve?

Issue Number: ref #29167 #29156 #29355 #29356 

Problem Summary:
Before test GBK support for `aes_decrypt` and `compress` , manually convert charset of args to GBK, this is due to misuse of `datumsToConstant`, which will not attach charset and collate info to args.
Test in `sha2`, `sha` also misuse `datumsToConstant`.

### What is changed and how it works?
- Add target charset info via manually set in test of `aes_decrypt` , use `primitiveValsToConstants` in test of `compress`, `sha2`, `sha`, which will attach charset and collate info.

### Check List

Tests

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by unconsolable** [comment]
Encounter common failures on `test session with real tikv exclude` and `test session with real tikv suites`, unit tests is fine.

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by zimulala** [comment]
PTAL @xiongjiwei @tangenta @Defined2014

File: expression/builtin_encryption_test.go [hunk 1]
```diff
@@ -227,7 +227,14 @@ func TestAESDecrypt(t *testing.T) {
...
 			args = append(args, types.NewDatum(param))
 		}
-		f, err := fc.getFunction(ctx, datumsToConstants(args))
+		// Set charset and collate for args to test new charset support
+		cons := datumsToConstants(args)
+		chs, col := ctx.GetSessionVars().GetCharsetInfo()
+		// Start from 1, as first arg is crypt_str
+		for i := 1; i < len(cons); i++ {
+			cons[i].GetType().Charset, cons[i].GetType().Collate = chs, col
+		}
```
Comment (Defined2014) [inline expression/builtin_encryption_test.go:236]: I think L226 - L236 could change to 
```
args := datumsToConstants([]types.Datum{fromHex(tt.crypt)})
args = append(args, primitiveValsToConstants(ctx, tt.params)...)
```

The `primitiveValsToConstants` func is similar as your code here.

Comment (unconsolable) [reply expression/builtin_encryption_test.go]: Resolved.

**Review by Defined2014 — APPROVED** [review]

File: expression/builtin_encryption_test.go [hunk 2]
```diff
@@ -310,8 +309,8 @@ var sha1Tests = []struct {
 	{mysql.DefaultCollationName, 1024, "128351137a9c47206c4507dcf2e6fbeeca3a9079"},
 	{mysql.DefaultCollationName, 123.45, "22f8b438ad7e89300b51d88684f3f0b9fa1d7a32"},
 	{"gbk", 123.45, "22f8b438ad7e89300b51d88684f3f0b9fa1d7a32"},
-	{"gbk", "一二三", "01c1743ce7a7e822454a659f659bad61375ff10c"},
-	{"gbk", "一二三123", "7c9a76465a02c41d377596431ef29418e2f6a72c"},
+	{"gbk", "一二三", "30cda4eed59a2ff592f2881f39d42fed6e10cad8"},
```
Comment (xiongjiwei) [inline expression/builtin_encryption_test.go:312]: why the test case is out of a test function. can you please move it into the test function?

Comment (unconsolable) [reply expression/builtin_encryption_test.go]: Resolved, also I found `sha2Tests` only used within `TestSha2Hash` and move in

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**